src/modules/job-exec/exec.c

```diff
@@ -163,12 +163,53 @@ static void output_cb (struct bulk_exec *exec, flux_subprocess_t *p,
                         len);
 }
 
+static void lost_shell_continuation (flux_future_t *f, void *arg)
+{
+    struct jobinfo *job = arg;
+    if (flux_future_get (f, NULL) < 0)
+        jobinfo_fatal_error (job, errno,
+                             "failed to notify job of lost shell");
+    flux_future_destroy (f);
+}
+
+static int lost_shell (struct jobinfo *job,
+                       int shell_rank,
+                       const char *hostname)
+{
+    flux_future_t *f;
+
+    /* Raise a non-fatal job exception */
+    jobinfo_raise (job,
+                   "node-failure",
+                   2,
+                   "%s on %s (shell rank %d)",
+                   "lost contact with job shell",
+                   hostname,
+                   shell_rank);
+
+    /* Also notify job shell rank 0 of exception
+     */
+    if (!(f = jobinfo_shell_rpc_pack (job,
+                                      "exception",
+                                      "{s:s s:i s:i}",
+                                      "type", "lost-shell",
+                                      "severity", 2,
+                                      "shell_rank", shell_rank)))
+            return -1;
+    if (flux_future_then (f, -1., lost_shell_continuation, job) < 0) {
+        flux_future_destroy (f);
+        return -1;
+    }
+    return 0;
+}
```

Comment:
instead of hard coded 2, perhaps define some "NON_FATAL_SEVERITY" macro or something?

Reply:
I'm not sure we've standardized where to put that? Perhaps the numbering should just follow syslog severities? That feels like a whole other issue/RFC etc though

Reply:
yeah, I don't think we do, but it just felt weird to see the `2` in there all alone without context (had to verify it was severity).  Perhaps just "int non_fatal_severity = 2;` in the function to give it context?

Reply:
Ok, good point. I've force pushed a change to define a `FLUX_JOB_EXCEPTION_CRIT` in that file to `2`. At least we'll be able to search the code for it if we ever define these in the API or somewhere more centralized.